### PR TITLE
Add command line options to select a single pass and to specify output files

### PR DESCRIPTION
**Pull request by nbp**
see https://bugzilla.mozilla.org/show_bug.cgi?id=1281572

### CODE REVIEW

File: iongraph [hunk 1]
```diff
@@ -255,6 +255,15 @@ def buildOnlyFinalPass(func):
...
     return [buildGraphsForPass(p)]
 
+# Write out a graph, in a given filename.
+def outputPassToFile(g, filename):
+    if filename == "-":
+        f = sys.stdout
+    else:
+        f = open(filename, 'w')
+    g.write(f)
+    f.close()
```
Comment (sstangl) [inline iongraph]: On one path, this closes stdout. That probably isn't expected behavior!

File: iongraph [hunk 2]
```diff
@@ -255,6 +255,15 @@ def buildOnlyFinalPass(func):
     p = func['passes'][-1]
     return [buildGraphsForPass(p)]
 
+# Write out a graph, in a given filename.
+def outputPassToFile(g, filename):
+    if filename == "-":
```
Comment (sstangl) [inline iongraph]: This is a little confusing. "-" always means stdin! Maybe this function would be better off taking an fd?

File: iongraph [hunk 3]
```diff
@@ -320,19 +327,86 @@ def main():
...
-            except e:
+            except:
+                dieWithUsage()
+            args.pop(i)
+            args.pop(i)
+            break
+
+    # --passnum N option: select the pass number, by index.
+    optPass = -1
+    for i in range(0, len(args)):
```
Comment (sstangl) [inline iongraph]: This file is python2, so it should use xrange for iteration to be idiomatic. I should convert it to python3 sometime...

File: iongraph [hunk 4]
```diff
@@ -320,19 +327,86 @@ def main():
         if args[i] == '--funcnum':
             try:
                 optFunc = int(args[i+1])
-            except e:
+            except:
+                dieWithUsage()
+            args.pop(i)
+            args.pop(i)
```
Comment (sstangl) [inline iongraph]: All the instances of args.pop(i) are incorrect. You can't modify an array under iteration. This is incorrect in the code as-is, but shouldn't be excerbated.

For example:
```
x = [1,2,3,4,5,6]
for k in xrange(0, len(x)):
    if k == 2:
        x.pop(1)
        x.pop(1)
```
Will visit 0,1,2,3,4,5. Not what's expected.

File: iongraph [hunk 5]
```diff
@@ -346,6 +420,22 @@ def main():
 
         gtl = buildOnlyFinalPass(func) if optFinal else buildGraphs(func)
 
+        if optName != "" and func['name'] != optName:
+            continue
```
Comment (sstangl) [inline iongraph]: I'm not a fan of this short-circuiting layout. This style is best used when the purpose of the function is to do one thing, and we're getting prerequisites out of the way. It doesn't belong in a function that does many things. Could it be rewritten as `if optName and func['name'] != optname:` ?

File: iongraph (hunk 2 above)
Comment (nbp) [inline iongraph]: I disagree, "-" means either stdin, or stdout, based on the context of the argument.

File: iongraph (hunk 1 above)
Comment (nbp) [inline iongraph]: That's totally expected.  If you ask to output on stdout, you probably don't want anything else to be spewed after your dot file.

File: iongraph (hunk 2 above)
Comment (nbp) [inline iongraph]: https://docs.python.org/dev/library/argparse.html
> FileType objects understand the pseudo-argument '-' and automatically convert this into sys.stdin for readable FileType objects and sys.stdout for writable FileType objects:

**Comment by nbp** [comment]
@sstangl , this new patch removes some of the boiler plate that I once added and never used since which is not required for calling iongraph within gdb.

File: iongraph [hunk 6]
```diff
@@ -327,6 +336,17 @@ def main():
             mir = gt[0]
             lir = gt[1]
 
+            sys.stderr.write(" found pass %d, expect pass %d.\n" % (j, args.passnum,))
```
Comment (sstangl) [inline iongraph]: This looks like debugging stuff!

File: iongraph [hunk 7]
```diff
@@ -256,17 +256,19 @@ def buildOnlyFinalPass(func):
...
     funcid = str(fnum).zfill(2)
     passid = str(pnum).zfill(2)
 
     filename = 'func%s-pass%s-%s-%s.gv' % (funcid, passid, g.name, str(irname))
-
-    f = open(filename, 'w')
-    g.write(f)
-    f.close()
+    with open(filename, 'w') as fd:
+        outputPassToFile(g, fd)
```
Comment (sstangl) [inline iongraph]: Probably just g.write(fd) would be fine.

**Comment by nbp** [comment]
@sstangl , nits applied.

**Pull request merged**